api/v1beta1/azuremachine_types.go

```diff
@@ -118,6 +118,12 @@ type AzureMachineSpec struct {
 	// SubnetName selects the Subnet where the VM will be placed
 	// +optional
 	SubnetName string `json:"subnetName,omitempty"`
+
+	// NetworkInterfaces specifies a list of network interface configurations.
+	// If left unspecified, the VM will get a single network interface with a
+	// single IPConfig in the subnet specified in the spec.subnetName field.
+	// +optional
+	NetworkInterfaces []NetworkInterface `json:"networkInterfaces,omitempty"`
```

Comment:
Thinking about this a bit more, what's potentially confusing here is that the naming and the comment of the field imply that this is defining _all_ network interfaces for that VM. That means as a user if I specify this field I would need to somehow specify the primary NIC and IPs as well. But looking at the code, it only is used for secondary NICs that get added on top of the default one (and that one is not configurable AFAICT). 

Would it make more sense to rename this to something like `SecondaryNetworkInterfaces` or `AdditionalNetworkInterfaces` and adjust the comment to make it match what the code it doing: adding these optional NICs to the VM if they are specified, in addition to the always present primary NIC.

Reply:
This field is actually defining all network interfaces for the VM. That is needed to specify the number of IPConfigs on the primary interface if you are using Azure CNI on the primary network interface.

Reply:
In that case it makes https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2411#discussion_r914260114 and https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2411#discussion_r1012367038 even more relevant